### ISSUE TITLE
Wrong error : Object is possibly 'undefined'. (TS2532)

### ISSUE THREAD

**Issue by hckhanh** [microsoft/TypeScript#41542]
**TypeScript Version:**  v4.2.0-dev.20201112

**Search Terms:** TS2532, undefined

**Code**

```ts
export default function App() {
  const { data, error } = useSWR<object, Error>(
    "https://jsonblob.com/api/jsonBlob/01861067-2595-11eb-bacf-717e52210c0a"
  );

  if (!data && !error) {
    return <div>Loading...</div>;
  } else if (error) {
    return <div>Error: {error.message}</div>;
  }

  return <div>{data.array}</div>;
}
```

**Expected behavior:** There is no way `data` can be `undefined` when it's rendered

**Actual behavior:** Object is possibly 'undefined'. (TS2532) appears at `data.array`

**Playground Link:** https://codesandbox.io/s/object-is-possibly-undefined-ts2532-n3kxp

**Related Issues:** I already searched but no result for me.

**Comment by RyanCavanaugh** [comment]
We don't have any mechanisms in place that can account for the complementarity of `x` and `!x` and don't plan to add any. There's no reason not to write this code more straightforwardly:
```ts
if (data) {
  return data.array;
}
if (error) {
  return "error";
}
return "loading";
```

**Comment by nhhockeyplayer** [comment]
this-issue is terrible

its wiped out my pre handler hook I cannot even reference THIS on any of these lines and its blocking
all attempts to augment any reference to THIS fail

```
@pre<IQuestion>('save', (next) => {
    console.log('HERE IT IS')
    console.log(this)

    if (this.createdAt !== null) {
        this.createdAt = this.updatedAt = new Date()
    }
    this.updatedAt = new Date()
    next()
})
export class Question extends Typegoose implements IQuestion {
```

**Comment by RyanCavanaugh** [comment]
@nhhockeyplayer can you please log a new issue with a complete repro? This doesn't look related to the problem described in the OP

**Comment by nhhockeyplayer** [comment]
Ok I will check my fire

I backed down to 4.0.5 and its happening on both

I have alot of code spread out across several projects

dont be alarmed if its legit you will hear more from me

Im still in shakedown mode

**Comment by icecream17** [comment]
I understand that there's no way to detect what the code does, but how can I change this code to make ts not throw this error?

```typescript
interface exampleType {
   property?: [number]
}

function example(someObject: exampleType) {
   if (!('property' in someObject)) someObject.property = [0]
   return someObject.property[0] // Error: Object is possibly undefined
}

example({})
```

**Comment by nhhockeyplayer** [comment]
this remains an issue

TS2532: Object is possibly 'undefined'.

```
@pre<RoleEntity>('save', (next) => {
    if (!this.createdAt) {
        this.createdAt = this.updatedAt = new Date()
    }
    this.updatedAt = new Date()
    next()
})
export class ...
```

**Comment by nhhockeyplayer** [comment]
its not very encouraging to close something that was not even vetted to be closed

does integrity exist in this world?

why would someone do this?

just to close it and make it gone? Does that do truth anything?
??

**Comment by icecream17** [comment]
Maybe try:
```ts
    this.createdAt ??= this.updatedAt = new Date()
```

**Comment by zz-james** [comment]
I think the problem is that arrow functions don't have a this try using an old school function definition?

> this remains an issue
> 
> TS2532: Object is possibly 'undefined'.
> 
> ```
> @pre<RoleEntity>('save', (next) => {
>     if (!this.createdAt) {
>         this.createdAt = this.updatedAt = new Date()
>     }
>     this.updatedAt = new Date()
>     next()
> })
> export class ...
> ```
> 
> unable to operate THIS pointer

**Comment by schoettler** [comment]
![image](https://user-images.githubusercontent.com/1758136/151046171-ab559b73-7dcc-4ca0-99b9-e8cdbe297365.png)

@RyanCavanaugh

**Comment by schoettler** [comment]
![image](https://user-images.githubusercontent.com/1758136/151048135-647a3b62-643b-4ea6-8cd6-a75488d8351f.png)

@RyanCavanaugh
